### PR TITLE
feat: add `@agoric/base-zone`

### PR DESCRIPTION
**Pull request by michaelfig**
refs: #8088

## Description

Split the parts of `@agoric/zone` which don't rely on `@agoric/vat-data` into a new package, `@agoric/base-zone`.

This should help resolve a package dependency cycle, and make it easier to move and rename just `@agoric/base-zone` to `@endo/zone` when we're ready.

### Security Considerations

n/a

### Scaling Considerations

n/a

### Documentation Considerations

n/a

### Testing Considerations

No special.

### Upgrade Considerations

Due to namespace conflicts, the singleton `heapZone` and `virtualZone` are no longer available.  Instead, use `makeHeapZone()` to create a fresh `heapZone` when needed (same for `makeVirtualZone()`).

### CODE REVIEW

File: packages/base-zone/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Base Zone Library
```
Comment (turadg) [inline packages/base-zone/README.md:1]: I was surprised it's not called endo-zone but I agree this is better. The plan to eventually move to Endo needn't be captured in SCM. There's sufficient motivation without it.

File: packages/base-zone/README.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+
+This library is used internally by [`@agoric/zone`](../zone/README.md); refer to it for more details.  Unless you are an author of a new Zone backing store type, or want to use `makeHeapZone` without introducing build dependencies on `@agoric/vat-data`, you should instead use `@agoric/zone`.
```
Comment (turadg) [inline packages/base-zone/README.md]: Consider to flip the ordering to begin with what the package offers. Then why you would or wouldn't want to use it directly.

File: packages/base-zone/jsconfig.json [hunk 3]
```diff
@@ -0,0 +1,19 @@
+// This file can contain .js-specific Typescript compiler config.
+{
+  "compilerOptions": {
```
Comment (turadg) [inline packages/base-zone/jsconfig.json]: Please inherit from the root config, which defines most of these already.

File: packages/base-zone/package.json [hunk 4]
```diff
@@ -0,0 +1,51 @@
+{
+  "name": "@agoric/base-zone",
+  "version": "1.0.0",
```
Comment (turadg) [inline packages/base-zone/package.json]: Please use a 0 major version to postpone the breaking change discipline required by semver.

File: packages/base-zone/package.json [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  "author": "Agoric",
+  "license": "Apache-2.0",
+  "dependencies": {
+    "@agoric/store": "^0.9.2",
+    "@endo/exo": "^0.2.3",
+    "@endo/far": "^0.2.19",
+    "@endo/pass-style": "^0.1.4",
+    "@endo/patterns": "^0.2.3"
+  },
+  "devDependencies": {
```
Comment (turadg) [inline packages/base-zone/package.json:35]: This will still have a cycle. Maybe the base-zone tests that need internal live there or in zone?

File: packages/base-zone/package.json [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    "test:c8": "c8 $C8_OPTIONS ava --config=ava-nesm.config.js",
+    "test:xs": "exit 0",
+    "lint-fix": "yarn lint:eslint --fix",
+    "lint": "run-s --continue-on-error lint:*",
+    "lint:types": "tsc -p jsconfig.json",
+    "lint:eslint": "eslint ."
+  },
+  "exports": {
+    ".": "./src/index.js",
+    "./heap.js": "./heap.js",
```
Comment (turadg) [inline packages/base-zone/package.json:22]: Why have root heap.js? Exporting src/heap.js has the same API.

If there's a good reason to have a root file for each export, why not do the same for index?

File: packages/base-zone/tools/testers.js [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+  });
+  t.is(rootZone.isStorable(nonPassable), false);
+  t.is(subZone.isStorable(123), true);
+  t.throws(() => rootZone.makeOnce('nonPassable', () => nonPassable), {
+    message: /is not storable/,
+  });
+};
+
+// CAUTION: Do not modify this list; it exists to ensure that future versions
+// of @agoric/zone are compatible with the baggage created by older versions,
```
Comment (turadg) [inline packages/base-zone/tools/testers.js:27]: Mixing concerns. This won't go to Endo/zone

File: packages/base-zone/package.json (hunk 5 above)
Comment (michaelfig) [reply packages/base-zone/package.json:35]: I chased down all the cycles from `@agoric/internal`.  Lerna now just complains about a single nested cycle (deploy-script-support -> vats).

File: packages/base-zone/package.json (hunk 6 above)
Comment (michaelfig) [reply packages/base-zone/package.json:22]: There is a reason (engines or bundlers that don't support `"exports"`), and "src/index.js" is not necessary because it already appears in `"main"` (which has wide support).

File: packages/base-zone/tools/testers.js (hunk 7 above)
Comment (michaelfig) [reply packages/base-zone/tools/testers.js:27]: Actually, it will for historical reasons.  Without this list, the heap zone won't reveal problems that would surface with durable zones, which is one of the main reasons to use zones in the first place.

Until we have an upgrade strategy to move to a different key name allocation policy, this "mixed concern" needs to be enforced.

Comment (turadg) [reply packages/base-zone/tools/testers.js:27]: Ah. When it moves I expect we'll reword it to emphasize Endo's independence.

**Review by turadg — APPROVED** [review]
So happy to see this.

One more change is to update this part of internal's README:
> It must be the lowest agoric-sdk package in any import graph. Therefore it must never depend on another agoric-sdk package. If there's a module in an another agoric-package that has no agoric-sdk dependencies, it can be moved into this package.

Now it has a chain of local dependencies: `base-zone` --> `store` --> `assert`. Each of those is destined for Endo so it's a good graph, but the doc needs updating. Feel free to add in your squash rebase force push.

Perhaps,
> It is meant to be a home for modules that have no Agoric-specific dependencies themselves. It does depend on a these other `@agoric` packages but they are all destined to migrate out of the repo,
> - `base-zone`
> - `store`
> - `assert`
> 
> This package may not take dependencies on any others in this repository.

File: packages/base-zone/package.json (hunk 5 above)
Comment (turadg) [reply packages/base-zone/package.json:35]: Solid progress!

I'm seeing four cycles left for future PRs
```
@agoric/deploy-script-support -> @agoric/vats -> @agoric/inter-protocol -> @agoric/deploy-script-support
@agoric/governance -> (nested cycle: @agoric/deploy-script-support -> @agoric/vats -> @agoric/inter-protocol -> @agoric/deploy-script-support) -> @agoric/governance
@agoric/smart-wallet -> (nested cycle: @agoric/governance -> (nested cycle: @agoric/deploy-script-support -> @agoric/vats -> @agoric/inter-protocol -> @agoric/deploy-script-support) -> @agoric/governance) -> @agoric/smart-wallet
(nested cycle: @agoric/smart-wallet -> (nested cycle: @agoric/governance -> (nested cycle: @agoric/deploy-script-support -> @agoric/vats -> @agoric/inter-protocol -> @agoric/deploy-script-support) -> @agoric/governance) -> @agoric/smart-wallet) -> (nested cycle: @agoric/smart-wallet -> (nested cycle: @agoric/governance -> (nested cycle: @agoric/deploy-script-support -> @agoric/vats -> @agoric/inter-protocol -> @agoric/deploy-script-support) -> @agoric/governance) -> @agoric/smart-wallet)
```

**Comment by dckc** [comment]
While doing some out-of-this-repo dapp development, I discovered that this is a breaking change: it renames `heapZone` to `makeHeapZone`.

The new name is better, but the old name should have been kept as an alias for compatibility.

_p.s. looks like they're not synonyms; `heapZone` is a zone, not a function to make one. still a breaking change._

**Comment by michaelfig** [comment]
> _p.s. looks like they're not synonyms; `heapZone` is a zone, not a function to make one. still a breaking change._

I'll at least add this note to the PR description.
